Mathlib/Topology/GPseudoMetric/Basic.lean

```diff
@@ -0,0 +1,328 @@
+/-
+Copyright (c) 2024 Edward van de Meent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Edward van de Meent
+-/
+import Mathlib.Algebra.CovariantAndContravariant
+import Mathlib.Algebra.NeZero
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+/-!
+# General Pseudo-Metrics
+
+In this file we introduce `GPseudoMetric`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering.
+
+## Main Definitions
+
+- `GPseudoMetric α β`: A structure containing a distance function on `α` with codomain `β`,
+which may be equal to 0 for non-equal elements. the distance function is 0 for equal elements,
+is commutative in its arguments, and satisifies the triangle inequality.
+- `GPseudoMetricClass α β`: The class of types of generic pseudo metrics on `α` to `β`.
+
+Additional useful definitions:
+
+- `ball gdist x δ`: The set of points with distance to x strictly less than δ
+- `closedBall gdist x δ`: The set of points with distance to x less than or equal to δ
+- `sphere gdist x δ`: The set of points with distance to x equal to δ
+
+-/
+
+open Set
+
+/-- Generic pseudo metrics
+
+A generic pseudo metric is a distance function `gdist : α → α → β`, which is zero for
+identical elements, for which the arguments commute, and for which the triangle
+inequality holds. As opposed to a classical pseudo metric, the codomain of this distance
+function is not necessarily ℝ (or ℝ≥0∞), and as a result does not endow α with a uniform
+space.
+-/
+structure GPseudoMetric
+    (α : Type*) (β : Type*) [LinearOrder β] [AddCommMonoid β]
+    where
+  /-- A distance function on α with values in β -/
+  toFun (x y : α):β
+  gdist_self : ∀ x : α, toFun x x = 0
+  comm' : ∀ x y : α, toFun x y = toFun y x
+  triangle' : ∀ x y z : α, toFun x z ≤ toFun x y + toFun y z
+
+variable {α β :Type*} [LinearOrder β] [AddCommMonoid β] [CovariantClass β β (.+.) (.≤.)]
```

Comment:
Here and elsewhere: please put spaces around the colon, and around binary operators (i.e., `+` and `≤`).

And a blank line after the `variable`, please.